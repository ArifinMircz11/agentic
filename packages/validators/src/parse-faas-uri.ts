// TODO: investigate this
/* eslint-disable security/detect-unsafe-regex */

import type { ParsedFaasIdentifier } from './types'

// namespace/projectName@deploymentHash/servicePath
// project@deploymentHash/servicePath
const projectDeploymentServiceRe =
  /^([a-zA-Z0-9-]{1,64}\/[a-z0-9-]{3,64})@([a-z0-9]{8})(\/[a-zA-Z0-9\-._~%!$&'()*+,;=:/]*)?$/

// namespace/projectName@version/servicePath
// project@version/servicePath
const projectVersionServiceRe =
  /^([a-zA-Z0-9-]{1,64}\/[a-z0-9-]{3,64})@([^/?@]+)(\/[a-zA-Z0-9\-._~%!$&'()*+,;=:/]*)?$/

// namespace/projectName/servicePath
// project/servicePath (latest version)
const projectServiceRe =
  /^([a-zA-Z0-9-]{1,64}\/[a-z0-9-]{3,64})(\/[a-zA-Z0-9\-._~%!$&'()*+,;=:/]*)?$/

export function parseFaasUri(uri: string): ParsedFaasIdentifier | undefined {
  const pdsMatch = uri.match(projectDeploymentServiceRe)

  if (pdsMatch) {
    const projectId = pdsMatch[1]!
    const deploymentHash = pdsMatch[2]!
    const servicePath = pdsMatch[3] || '/'

    return {
      projectId,
      deploymentHash,
      servicePath,
      deploymentId: `${projectId}@${deploymentHash}`
    }
  }

  const pvsMatch = uri.match(projectVersionServiceRe)

  if (pvsMatch) {
    return {
      projectId: pvsMatch[1]!,
      version: pvsMatch[2]!,
      servicePath: pvsMatch[3] || '/'
    }
  }

  const psMatch = uri.match(projectServiceRe)

  if (psMatch) {
    return {
      projectId: psMatch[1]!,
      servicePath: psMatch[2] || '/',
      version: 'latest'
    }
  }

  // Invalid FaaS uri, so return undefined
  return
}
