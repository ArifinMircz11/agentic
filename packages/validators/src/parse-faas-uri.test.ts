import { expect, test } from 'vitest'

import { parseFaasUri } from './parse-faas-uri'

function success(value: string) {
  const result = parseFaasUri(value)
  expect(result).toBeTruthy()
  expect(result?.projectId).toBeTruthy()
  expect(result?.version || result?.deploymentHash).toBeTruthy()
  expect(result).toMatchSnapshot()
}

function error(value: string) {
  const result = parseFaasUri(value)
  expect(result).toBeUndefined()
}

test('username/projectName@deployment/servicePath success', () => {
  success('username/foo-bar@01234567/foo')
  success('username/foo-bar@abc123lz/foo')
  success('username/foobar123-yo@01234567/foo_bar_BAR_901')
  success('username/foobar@01234567/foo/bar/123/456')
})

test('username/projectName@deployment/servicePath error', () => {
  error('foo-bar@01234567/foo')
  error('%/foo-bar@01234567/foo')
  error('user/foo^bar@01234567/foo')
  error('user@foo^bar@01234567/foo')
  error('username/Foo-Bar@01234567/foo')
})

test('username/projectName@version/servicePath success', () => {
  success('username/foo-bar@latest/foo')
  success('username/foo-bar@dev/foo')
  success('username/foo-bar@1.0.0/foo')
  success('username/foobar123-yo@0.0.1/foo_bar_BAR_901')
  success('username/foobar123-yo@0.0.1/foo/bar/123-456')
})

test('username/projectName@version/servicePath error', () => {
  error('foo_bar@latest/foo')
  error('username/foo-bar@1.0.0/foo@')
  error('username/foo-bar@/foo')
  error('username/foo-bar@/foo/')
  error('username/fooBar123-yo@0.0.1/foo/bar/123-456')
})

test('username/projectName/servicePath success', () => {
  success('u/foo-bar/foo')
  success('a/foo-bar/foo_123')
  success('foo/foobar123-yo/foo_bar_BAR_901')
  success('foo/foobar123-yo/foo/bar/123/456')
})

test('username/projectName/servicePath error', () => {
  error('@/foo_bar/foo')
  error('foo-bar/foo\\/')
  error('user/_/foo')
  error('user/a 1/foo')
  error('u/FOO-bar/foo')
})

test('username/projectName@deployment success', () => {
  success('abc/hello-world@3d2e0fd5')
  success('a16z/foo-bar@f673db32c')
  success('foodoo/foo-bar@f673db32c')
  success('u/foobar123-yo@673db32c')
  success('username/foo-bar@01234567/')
})

test('username/projectName@deployment error', () => {
  error('/hello-world@3d2e0fd5')
  error('foo-bar@f673db32c')
  error('foodoo/foo@bar@f673db32c')
  error('u/fooBar123-yo@/673db32c')
  error('abc/Hello-World@3d2e0fd5')
})

test('username/projectName@version success', () => {
  success('abc/hello-world@1.0.3')
  success('a16z/foo-bar@latest')
  success('a16z/foo-bar@dev')
  success('foodoo/foo-bar@1.0.1')
  success('u/foobar123-yo@3.2.2234')
  success('username/foo-bar@1.0.3/')
})

test('username/projectName@version error', () => {
  error('/hello-world@3d2e0fd5')
  error('foo-bar@f673db32c')
  error('foodoo/foo@bar@f673db32c@')
  error('u/fooBar123-yo@/673db32c/')
  error('abc/hello-World@1.0.3')
})

test('username/projectName success', () => {
  success('abc/hello-world')
  success('a16z/foo-bar')
  success('foodoo/foo-bar')
  success('u/foobar123-yo')
  success('abc/hello-world/')
})

test('username/projectName error', () => {
  error('/hello-world')
  error('foo-barc')
  error('foodoo/foo@bar@')
  error('u/fooBar123-yo@/')
  error('abc/HELLO-WORLD')
})
