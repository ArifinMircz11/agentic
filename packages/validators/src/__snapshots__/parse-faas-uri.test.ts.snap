// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`username/projectName success 1`] = `
{
  "projectId": "abc/hello-world",
  "servicePath": "/",
  "version": "latest",
}
`;

exports[`username/projectName success 2`] = `
{
  "projectId": "a16z/foo-bar",
  "servicePath": "/",
  "version": "latest",
}
`;

exports[`username/projectName success 3`] = `
{
  "projectId": "foodoo/foo-bar",
  "servicePath": "/",
  "version": "latest",
}
`;

exports[`username/projectName success 4`] = `
{
  "projectId": "u/foobar123-yo",
  "servicePath": "/",
  "version": "latest",
}
`;

exports[`username/projectName success 5`] = `
{
  "projectId": "abc/hello-world",
  "servicePath": "/",
  "version": "latest",
}
`;

exports[`username/projectName/servicePath success 1`] = `
{
  "projectId": "u/foo-bar",
  "servicePath": "/foo",
  "version": "latest",
}
`;

exports[`username/projectName/servicePath success 2`] = `
{
  "projectId": "a/foo-bar",
  "servicePath": "/foo_123",
  "version": "latest",
}
`;

exports[`username/projectName/servicePath success 3`] = `
{
  "projectId": "foo/foobar123-yo",
  "servicePath": "/foo_bar_BAR_901",
  "version": "latest",
}
`;

exports[`username/projectName/servicePath success 4`] = `
{
  "projectId": "foo/foobar123-yo",
  "servicePath": "/foo/bar/123/456",
  "version": "latest",
}
`;

exports[`username/projectName@deployment success 1`] = `
{
  "deploymentHash": "3d2e0fd5",
  "deploymentId": "abc/hello-world@3d2e0fd5",
  "projectId": "abc/hello-world",
  "servicePath": "/",
}
`;

exports[`username/projectName@deployment success 2`] = `
{
  "projectId": "a16z/foo-bar",
  "servicePath": "/",
  "version": "f673db32c",
}
`;

exports[`username/projectName@deployment success 3`] = `
{
  "projectId": "foodoo/foo-bar",
  "servicePath": "/",
  "version": "f673db32c",
}
`;

exports[`username/projectName@deployment success 4`] = `
{
  "deploymentHash": "673db32c",
  "deploymentId": "u/foobar123-yo@673db32c",
  "projectId": "u/foobar123-yo",
  "servicePath": "/",
}
`;

exports[`username/projectName@deployment success 5`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/",
}
`;

exports[`username/projectName@deployment/servicePath success 1`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foo-bar@01234567",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`username/projectName@deployment/servicePath success 2`] = `
{
  "deploymentHash": "abc123lz",
  "deploymentId": "username/foo-bar@abc123lz",
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
}
`;

exports[`username/projectName@deployment/servicePath success 3`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foobar123-yo@01234567",
  "projectId": "username/foobar123-yo",
  "servicePath": "/foo_bar_BAR_901",
}
`;

exports[`username/projectName@deployment/servicePath success 4`] = `
{
  "deploymentHash": "01234567",
  "deploymentId": "username/foobar@01234567",
  "projectId": "username/foobar",
  "servicePath": "/foo/bar/123/456",
}
`;

exports[`username/projectName@version success 1`] = `
{
  "projectId": "abc/hello-world",
  "servicePath": "/",
  "version": "1.0.3",
}
`;

exports[`username/projectName@version success 2`] = `
{
  "projectId": "a16z/foo-bar",
  "servicePath": "/",
  "version": "latest",
}
`;

exports[`username/projectName@version success 3`] = `
{
  "projectId": "a16z/foo-bar",
  "servicePath": "/",
  "version": "dev",
}
`;

exports[`username/projectName@version success 4`] = `
{
  "projectId": "foodoo/foo-bar",
  "servicePath": "/",
  "version": "1.0.1",
}
`;

exports[`username/projectName@version success 5`] = `
{
  "projectId": "u/foobar123-yo",
  "servicePath": "/",
  "version": "3.2.2234",
}
`;

exports[`username/projectName@version success 6`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/",
  "version": "1.0.3",
}
`;

exports[`username/projectName@version/servicePath success 1`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
  "version": "latest",
}
`;

exports[`username/projectName@version/servicePath success 2`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
  "version": "dev",
}
`;

exports[`username/projectName@version/servicePath success 3`] = `
{
  "projectId": "username/foo-bar",
  "servicePath": "/foo",
  "version": "1.0.0",
}
`;

exports[`username/projectName@version/servicePath success 4`] = `
{
  "projectId": "username/foobar123-yo",
  "servicePath": "/foo_bar_BAR_901",
  "version": "0.0.1",
}
`;

exports[`username/projectName@version/servicePath success 5`] = `
{
  "projectId": "username/foobar123-yo",
  "servicePath": "/foo/bar/123-456",
  "version": "0.0.1",
}
`;
